### PR TITLE
[WIP] Added new toolchain-based Meson helper

### LINKED ISSUE #7026: [question\rfc] meson: environment variable usage rework
Current `meson` integration uses the following approach:
- Generate environment via `AutoToolsBuildEnvironment`.
- Use env_vars (which sets up `CXXFLAGS`, `LDFLAGS` and etc) from this environment when executing all `ninja`/`meson` commands.

There are two issues(?) with such approach:
1. All dependency information (include paths, library search paths, defines and etc) is passed via global environment variables (e.g. `CPPFLAGS`), which means they apply to all the code in the package. Which result in a lot of extra paths to search, e.g. a case when a recipe has a lot of deps, but each library in this recipe only has a single dep. This also results in conflicts sometime: e.g. dep's headers are included with `-I`, but the project tries to include them with `-isystem` which might or might not work, depending on compiler and warning flags.
2. `meson` greatly discourages the use of environment variables and has explicit command line arguments to specify most of the flags instead. Current situation results in a weird situations, e.g. `fPIC` specified both in env vars and via command line. `meson` also might ignore some flags from env vars altogether.

Both problems can be fixed by not using `environment_append` in [this code](https://github.com/conan-io/conan/blob/02d9775c1c8dcd5580984ebf9a4fb638cbc1fd7a/conans/client/build/meson.py#L171) and by:

Problem 1.: using deps in `meson` through `pkg-config` generator (which is the recommended `mesonic` way).
Problem 2.: passing the arguments manually to `meson` during configuration.

Since this will be a breaking change, the easiest way would be to add a flag to `conans.Meson` ctor, e.g `Meson(generate_env=True)`, so by default the behaviour would remain the same, but the user could opt-in for the *clean* approach.

Note: this does approach does not aim to disable the whole env vars usage (e.g. from profiles), just the ones that are formed dynamically from `conan` settings and recipe deps.

I'm willing to try my hand at such PR, if such approach is approved.

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

### PR DESCRIPTION
**Pull request by TheQwertiest**
Changelog: (Feature | Fix | Bugfix): Describe here your pull request
Docs: **TODO** <s>https://github.com/conan-io/docs/pull/XXXX</s>
Fixes: https://github.com/conan-io/conan/issues/7026

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

Differences from stock Meson integration:
- does not override `default_library` if no `shared` options was set
- saves `build_subdir` in `__init__`
- cleaned up interface: removed old arguments, removed build_dir and etc
- added machine file support
- added machine file generation for `toolchain`
- backend-agnostic meson methods are used by default with ninja as a fallback for older meson versions.
- uses only command options to pass info to meson during configuration.
- env variables with dependency search paths are not appended when executing meson commands
- requires pkg-config generator
- install_folder is always added to pkg-config search paths now (since pc generator is required now)

Limitations:
Can't read `--profile:build`, thus works only in cases when `native` compiler is not required by meson project.

TODO:
- [ ] fix env variables reading for native\cross once it's possible (i.e when <https://github.com/conan-io/conan/issues/7091> is fixed)
- [ ] add pkg_config_path to configure properly (cross scenario and native) or move it to machine file generation
- [ ] cleanup comments
- [ ] add tests
- [x] deal with typings (remove)

Questions:
- <s>Can we use Py3.x? meson requires Python 3.5+</s> Can use 3.5, but it should not leak in global scope.
- Should we keep `cross-file`/`native-file` kwarg or should they be moved to `options` instead?
- Same question as ^ for pkg-config-paths
- How should the user extend `_get_cpu_family_and_endianness_from_arch` in case he has a custom unknown arch?
- Should we generate machine files in `dump()` or in `__init__()`?
- Move `MesonDefaultToolchainGenerator` to `MesonDefaultToolchain`?

### CODE REVIEW

**Comment by TheQwertiest** [comment]
CC @jgsogo, @KristianJerpetjon

**Comment by TheQwertiest** [comment]
This implementation supports the following cases:
- No profile (everything set via `settings`)
  - Native build.
  - Cross build (i.e. specified `arch` is different from native).
- Single `--profile`:
  - Profile contains only cross info.
  - Profile contains cross and native info (via `ENV_FOR_BUILD` + `settings.*_build`).
- Two profiles (`--profile:build --profile:host`): it will work, but only if information from `--profile:build` is not required by meson project ( #7091 ).

File: conans/client/build/mesonx.py [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+        build_dir = self._conanfile.build_folder
+        if self._build_subdir:
+            build_dir = str(Path(self._conanfile.build_folder) / _build_subdir)
+        return (source_dir, build_dir)
+
+    def _get_default_options(self) -> dict:
+        options = dict()
+        if self._conanfile.package_folder:
+            options['prefix'] = self._conanfile.package_folder
+        options['libdir'] = DEFAULT_LIB
```
Comment (memsharded) [inline conans/client/build/mesonx.py]: This is a bit dirty, we need a better way to specify the layout, this is hardcoded, even if the ``package_info()`` defines different directories, this will package to a different place, which will be confusing.

I'd prefer to fail at the moment, until we come up with a better way.

File: conans/client/build/mesonx.py [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+            options['cpp_std'] = self._get_meson_cppstd(cppstd)
+
+        shared = self._so('shared')
+        if shared != None:
+            options['default_library'] = 'shared' if shared else 'static'
+
+        host_os = self._ss('os')
+        if host_os and 'Windows' in host_os:
+            self._conanfile.output.warn("Toolchain: Ignoring fPIC option defined for Windows")
+        else:
```
Comment (memsharded) [inline conans/client/build/mesonx.py:270]: Shouldn't these things be managed in the toolchain files, and not in the build-helper?

File: conans/client/build/mesonx.py [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+    @staticmethod
+    def _get_meson_cppstd(cppstd):
+        cppstd_conan2meson = {
+            '98': 'c++03', 'gnu98': 'gnu++03',
+            '11': 'c++11', 'gnu11': 'gnu++11',
+            '14': 'c++14', 'gnu14': 'gnu++14',
+            '17': 'c++17', 'gnu17': 'gnu++17',
+            '20': 'c++1z', 'gnu20': 'gnu++1z'
+        }
+        if cppstd not in cppstd_conan2meson:
```
Comment (memsharded) [inline conans/client/build/mesonx.py:307]: Shouldn't this be managed in the toolchain, not here in the build-helper?

File: conans/client/toolchain/meson.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+
+        if path and config:
+            raise ConanException('Both `path` and `config` were supplied: only one should be used')
+        if path:
+            config = ConfigParser()
+            config.read(path)
+        self.options = config
+
+    def dump(self, install_dir: str):
+        outpath = Path(install_dir)
```
Comment (memsharded) [inline conans/client/toolchain/meson.py:30]: Probably not worth supporting Python27 at this point, but ``Pathlib`` hasn't been introduced yet into the conan codebase, probably not introduce it for consistency.

File: conans/client/build/mesonx.py (hunk 2 above)
Comment (TheQwertiest) [reply conans/client/build/mesonx.py:270]: Meson machine files are (mostly) for describing the machine environent (hence the name), i.e. architecture, compiler, search paths and etc. Options like `fPIC` and `std=c++**` are configuration options (e.g compiler usually supports more than one `std` mode and different projects might use different values even if used with the same machine file) and can't be specified in said files.

File: conans/client/build/mesonx.py (hunk 3 above)
Comment (TheQwertiest) [reply conans/client/build/mesonx.py:307]: See above.

File: conans/client/toolchain/meson.py (hunk 4 above)
Comment (TheQwertiest) [reply conans/client/toolchain/meson.py:30]: Sorry, I'm getting confused here: should I rewrite the code to use only Python 2.7? And if not and using Python 3.5+ is alright, then what would be the problem with `Pathlib` introduction? It's a (well-tested) part of standard library (and available since Python 3.4). I mean it's like using old Python string formatting syntax(`%s %s` and friends) when `'{}'.format(arg)` or `f'{arg}'` is available, i.e. it will make code only harder to read and more bug-prone.

File: conans/client/build/mesonx.py (hunk 1 above)
Comment (TheQwertiest) [reply conans/client/build/mesonx.py]: Uhm, but won't this fail during local dev flow? For example, a user invokes `conan configure . -bf=local_folder -sf=some_folder`, but, since `package_folder` is not available, this command will raise.

**Comment by TheQwertiest** [comment]
@memsharded , @jgsogo , I'd really like to move forward with this PR, esp. in the light of the following quote by jgsogo:

> 1.27 is almost complete and we need to start merging the PRs that are already opened and there are still some issues pending

I've replied to all the raised issues in this PR. But I can't do much without further input from conan devs (except for adding more tests).

File: conans/client/build/mesonx.py [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+
+# Questions:
+# - Should we keep `cross-file`/`native-file` kwarg or should they be moved to `options` instead?
+# - Same q as ^ for pkg-config-paths
+# - How should the user extend `_get_cpu_family_and_endianness_from_arch` in case he has a custom unknown arch?
+# - Should we generate machine files in dump() or in __init__()?
+# - Move `MesonDefaultToolchainGenerator` to `MesonDefaultToolchain`?
+
+class MesonX():
+    def __init__(self, conanfile, build_subdir: str = None, backend: str = None, append_vcvars: bool = False) -> None:
```
Comment (memsharded) [inline conans/client/build/mesonx.py]: Please remove all type annotations. We are not using them yet, and they are completely breaking Python2.7 clients. It is fine if the inner functionality only works with Python 3, but the introduction of this cannot break legacy Python 2.7, even if they are not using Meson: https://conan-ci.jfrog.info/blue/organizations/jenkins/ConanTestSuite/detail/PR-7155/2/pipeline

CI should be green, the tests using this functionality could be using ``@unittest.skipIf(six.PY2,...``

File: conans/client/build/mesonx.py (hunk 1 above)
Comment (memsharded) [reply conans/client/build/mesonx.py]: It is better to provide partial functionality at the moment and raise if something is not implemented yet (for example, the local flow), than provide flaky or hardcoded assumptions, like using DEFAULT_BIN and DEFAULT_LIB directories, that some users will start relying on that behavior, and later will be broken when we change it.

File: conans/client/build/mesonx.py [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+        shared = self._so('shared')
+        if shared != None:
+            options['default_library'] = 'shared' if shared else 'static'
+
+        host_os = self._ss('os')
+        if host_os and 'Windows' in host_os:
+            self._conanfile.output.warn("Toolchain: Ignoring fPIC option defined for Windows")
+        else:
+            fpic = self._so('fPIC')
+            if fpic != None:
```
Comment (memsharded) [inline conans/client/build/mesonx.py]: ``fpic is not None``

File: conans/client/build/mesonx.py (hunk 2 above)
Comment (memsharded) [reply conans/client/build/mesonx.py:270]: The point of the ``toolchain()`` feature is that for users working locally, they can build with their command line of the build system, and obtain the same result as if building with the ``build()`` method in the Conan cache. So having a very long command line that needs to include all the options, like:

```bash
$ meson -Db_staticpic=True -Db_ndebug=if-release -Dcpp_std=...
```
If all those inputs were already introduced by Conan in the previous ``conan install`` command, the idea of the ``toolchain()`` is that they are saved in a file or something that the build system and the users can easily use.

**Review by memsharded — COMMENTED** [review]
The main idea of the ``toolchain()`` approach is that it captures information passed at ``conan install`` time, and put it in a build-system format, so calling the build system manually is both simple and also achieve exactly the same build that with ``build()`` method and ``conan create``. 

So it excedes a bit every build system own definition of "toolchain". The Conan ``toolchain()`` could also generate virtualenv scripts if it is necessary to achieve a correct build that is using tools stored in Conan packages. Managing the environment is on the TODO list.

I would say that the current approach is very restricted to the Meson definition of toolchain, and thus the new new ``MesonX`` build system helper is still doing too much logic, so it might be missing something.

Also, I am fine if the helper and toolchain only works in Python 3.5+, but they cannot break Python 2.7 users that are not using it. So type annotations needs to be removed. In the same way module scope imports of ``from pathlib import Path`` are not possible.

Asking for @jgsogo feedback on this PR.

File: conans/client/build/mesonx.py (hunk 5 above)
Comment (TheQwertiest) [reply conans/client/build/mesonx.py]: I see, will be fixed. I'm a Python novice so this is all new to me =)

File: conans/client/build/mesonx.py (hunk 1 above)
Comment (TheQwertiest) [reply conans/client/build/mesonx.py]: The problem is that nobody will use (and test) this feature if it's doesn't work in the very basic and common scenario (esp. when the current helper does provide such functionality). Maybe add a note that "local development flow is highly experimental and might be/will be changed in the future updates"?
^ this about about `prefix` only. I don't have much objections to removing `DEFAULT_BIN` and friends (I don't see why it's bad though, esp. since conan itself expects a fixed layout when searching for dep binaries).

File: conans/client/build/mesonx.py (hunk 2 above)
Comment (TheQwertiest) [reply conans/client/build/mesonx.py:270]: Hm... One alternative would be to save all options that don't fit in Meson's machine files in a separate file which is generated during `toolchain()` and parsed during `configure()`.

File: conans/client/build/mesonx.py (hunk 5 above)
Comment (TheQwertiest) [reply conans/client/build/mesonx.py]: Fixed.

File: conans/client/build/mesonx.py (hunk 1 above)
Comment (TheQwertiest) [reply conans/client/build/mesonx.py]: Note: removed DEFAULT_BIN and friends, but kept `prefix` for now.

File: conans/client/build/mesonx.py (hunk 6 above)
Comment (TheQwertiest) [reply conans/client/build/mesonx.py]: Fixed

**Comment by TheQwertiest** [comment]
PS: Not sure how to fix CI failure - it seems that it can't find `meson` binary on Macos and Windows agents. Is there a way to require python packages from tests?

**Comment by jgsogo** [comment]
> PS: Not sure how to fix CI failure - it seems that it can't find `meson` binary on Macos and Windows agents. Is there a way to require python packages from tests?

Yes, it looks like `Meson` is not installed in those agents. We really need Meson in order to properly test this _toolchain_, no more mocking indeed.

---

About the PR. I'm having a look to the Meson documentation and I'm not able to find a way to resemble a CMake toolchain. There is this (https://github.com/mesonbuild/meson/issues/2472) issue opened asking for it, but from my understanding the only way to provide many inputs is via environment variables (linux-style) or using [properties in the native-file](https://mesonbuild.com/Machine-files.html) that requires the build script to take them into account using `meson.get_external_property` or `meson.get_cross_property`, something that [couples the build-script to the _toolchain file_](https://github.com/mesonbuild/meson/issues/2472#issuecomment-335823801) which is not the way to go for compiler or linker flags.

This section of the docs was promising (https://mesonbuild.com/Native-environments.html), but I end up in the same alternatives listed before.

Having a look to the environment way, and from the issue linked above, one arrives at [this line of code](https://github.com/mesonbuild/meson/blob/master/mesonbuild/compilers/compilers.py#L915), the function `get_args_from_envvars` and, the `get_global_options` as well, it looks like it is the way used by Meson to populate flags and options and write them in generated build-files.

IMO, **the _toolchain_ for Meson should follow the environment approach** and add a _native-file_ in order to provide other `key=value` pairs that the `meson.build` should retrieve using the `get_xxxx_property` I mentioned above. Is this how people provide values (like CMake definitions) to Meson? Is there any other way?

Indeed, I see this PR is using the envvars approach before running `meson setup`:

```python
        env_vars_to_clean = {
            'CC',
            'CXX',
            'LD',
            'CCFLAGS',
            'CXXFLAGS',
            'CPPFLAGS',
            'LDFLAGS',
            'AR',
            'AS',
            'STRIP',
            'RANLIB',
        }
        clean_env = {ev: None for ev in env_vars_to_clean}
        clean_env.update({'{}_FOR_BUILD'.format(ev): None for ev in env_vars_to_clean})

        with environment_append(clean_env):
            self._run('meson setup "{}" "{}" {}'.format(source_dir, build_dir, arg_list))
```

---

As @memsharded mentioned above, the environment is also needed to have full integration with CMake, so it is something in the roadmap. Meson toolchain (at least for the current release) should follow this same approach, I see no other way right now.

**Comment by TheQwertiest** [comment]
> IMO, the toolchain for Meson should follow the environment approach and add a native-file in order to provide other key=value pairs that the meson.build should retrieve using the get_xxxx_property I mentioned above. Is this how people provide values (like CMake definitions) to Meson? Is there any other way?

Currently (it might be changed pretty soon though with the following PR: https://github.com/mesonbuild/meson/pull/6597), the only mesonic way is the following:
- Describe machine environment in machine files (whether it's build machine or host machine).
- Set meson options (aka built-in options) and project options (aka user options) via command line. Machine files there are passed via cmdline as well.

`get_xxxx_property` should never be used for storing random options and was intended for storing various machine-specific data (e.g. `sound_device_id = FTX27.1`) that should not be exposed via project options.

Using env variables is considered harmful and not mesonic and is left there for compatibility only.

[EDIT]

> I'm having a look to the Meson documentation and I'm not able to find a way to resemble a CMake toolchain

I don't think it's good to think of `toolchain()` only in a scope of CMakes' toolchain. `toolchain()` is a generic, build-system agnostic method that must have a strict and build-system agnostic purpose or problem that it's designed to solve. Forcing every build system into behaving like CMake is wrong, inefficient and just not always possible.

[EDIT_END]

So, what do we have left then?

Well, first, I need to understand what do you consider as `toolchain()` and what do you expect from it. Cmake helper PR and linked issue had pretty rough and uninformative description (probably because you discussed it mostly internally, outside of GH), so I've implemented this method only using my common sense (which might be lacking :D). Hence why this PR was marked as WIP.

So, I'll describe what is *my* interpretation of a generic `toolchain()` method (based on https://github.com/conan-io/conan/pull/7155#discussion_r446963806). And I have to apologize in advance - English is not my native language, thus we might have a miscommunication every now and then.

#### `toolchain()`:

Prerequisites: a project that uses a two-step build system, meaning `configure [ARGS]` and `build`.

This method generates all args needed for the invocation of `configure` and stores it for a later use in said `configure` step. The way that these args are stored or passed from the storage to `configure` later is implementation defined. If this method is used then no additional arguments or options should be internally passed to `configure` unless they can't be divined during `toolchain()` call (e.g. `prefix` or `builddir`). Part I'm unsure of: this limitation does?/doesn't? apply to user options and overrides.

**Comment by TheQwertiest** [comment]
And as I've mentioned before, I'm available on conan's Slack, which might be a better venue for such long conversations =)

**Comment by TheQwertiest** [comment]
Added a [EDIT] section to my (long?) answer above.

**Comment by TheQwertiest** [comment]
Re-added `meson compile TARGET` functionality, since the relevant PR was merged in meson.

**Comment by jgsogo** [comment]
Wow! That PR is really interesting, I see some overlapping between that implementation and Conan as a package manager, but nothing we cannot fit. Most of us are not English speakers as well, so I hope we manage to understand each other 😄 

Our _toolchain_ file is something that let us work as follows:

 1. Only one call to Conan should be needed:

    ```
    conan install <path/to/conanfile.py> --profile=myprofile
    ```

    A file `conan_toolchain.meson` (name TBD, maybe it is the native-file) is generated, it contains everything related to settings and user definitions for the project: generator to use (visual studio version, ninja, Xcode,...), path to tools (compiler, linker,...), linker and compiler flags, user definitions,...

 2. From now on, only `meson` should be needed, and the only file needed to reproduce the binary that `conan create` would have built, should be the file `conan_toolchain.meson`. The user should run something like this:

    ```
    meson setup <build/dir> --toolchain=conan_toolchain.meson 
    meson compile -C <build/dir>
    ```

As you can see it is more than a _toolchain_, it not only contains compiler and linker related inputs, but also variables that the user might define in the conanfile:

```python
class Recipe:
    def toolchain(self):
        tc = MesonToolchain(self)
        tc.definitions["USER_INPUT"] = "value"
        tc.definitions.release["ONLY_FOR_RELEASE"] = "value"
        return tc
```

I would expect something like this for the generated file

**conan_toolchain.meson**/**native-file** (for a `Release` build)
```ini
[meson options]
# Related to env
env_path = <add/to/path>
env_variable = env_variable_value
# Related to settings
generator = vs2017
c_std = 'c14'
cpp_std = 'c++14'
c_args = ['-foo']
c_link_args = ['-Wl,--arg']
# Related to options (for the package being built)
default_library = 'static'
# User defined inputs
b_user_input = value
b_only_for_release = value
```

Of course, a lot of imagination is needed here 🧚, filenames and contents are totally speculative.

**Comment by jgsogo** [comment]
btw, I totally agree with this, but talking in terms of CMake offers something that is working right now, something concrete, and it is useful as a reference... but yes, CMake is not perfect, if we can improve, let's do it.

> I don't think it's good to think of toolchain() only in a scope of CMakes' toolchain. toolchain() is a generic, build-system agnostic method that must have a strict and build-system agnostic purpose or problem that it's designed to solve. Forcing every build system into behaving like CMake is wrong, inefficient and just not always possible.

**Comment by TheQwertiest** [comment]
Ok, we are getting somewhere =)

> From now on, only meson should be needed, and the only file needed to reproduce the binary that conan create would have built, should be the file conan_toolchain.meson

Well, that's currently impossible. Two machine files will be needed at the very least: native (aka build) and cross (aka host). And that's not counting user machine files.

And, as I've mentioned above, only a very limited number of options can be (currently) passed via machine files.

The easiest (and the dirtiest) way would be just to store the whole generated cmd line in a file (let's say `meson_setup_args.txt`). And then user could just call smth like ``meson setup `cat meson_setup_args.txt` ``.

> talking in terms of CMake offers something that is working right now, something concrete, and it is useful as a reference

It's only useful to those who are familiar with CMake and it's toolchain :P 
Of which I am not...

**Comment by jgsogo** [comment]
> Ok, we are getting somewhere =)
> 
> > From now on, only meson should be needed, and the only file needed to reproduce the binary that conan create would have built, should be the file conan_toolchain.meson
> 
> Well, that's currently impossible. Two machine files will be needed at the very least: native (aka build) and cross (aka host). And that's not counting user machine files.
> 

Ok, it can more than one file... especially if Meson users are already using several files. The idea is not to pass all the arguments one by one in the command line. For example, in the CMake use-case, these are all the command line inputs that are substituted by a single file:

image

image



> And, as I've mentioned above, only a very limited number of options can be (currently) passed via machine files.
> 

Probably this is the key point: **If it is how it works now (release 0.5x) or it is a design decision and it won't ever be changed.** 

⬆️ ⬆️ ⬆️ ⬆️ 

---

> The easiest (and the dirtiest) way would be just to store the whole generated cmd line in a file (let's say `meson_setup_args.txt`). And then user will just call smth like `` meson setup `cat meson_setup_args.txt` ``.
> 

I hope not, I prefer the environment variables approach... Conan can generate the corresponding `activate` and `deactivate` scripts.

> > talking in terms of CMake offers something that is working right now, something concrete, and it is useful as a reference
> 
> It's only useful to those who are familiar with CMake and it's toolchain :P
> Of which I am not...

touché! You win this time 😄

**Comment by TheQwertiest** [comment]
> If it is how it works now (release 0.5x) or it is a design decision and it won't ever be changed.

AFAIK most arguments could be specified once the mentioned PR is merged (and, as far as I can see, main meson maintainer/author is not against it). 
But machine file handling is needed even in the current (and older) meson versions as well. They are required for proper cross-builds, they are much better than having various env vars laying around, and it will also allow to invoke `meson` from command line directly (even if said command line would be long).
The only difference between *full-toolchain* (PR) and *partial-toolchain* (current status) would be the stuff from `configure()` method (i.e. the one from this PR).

TLDR: 
- current meson integration does not allow a build reproduction by copy-pasting cmd invocation (no matter how long it is), because of the env vars. (it does not allow a cross build either :D)
- machine file generation and usage should not be restricted to the latest meson version.
- partial toolchain-ing is better than the current status-quo and allows for cmd build reproduction (which is one of the goals of `toolchain()`).

**Comment by TheQwertiest** [comment]
@ milestone change. I don't think it'll be possible to complete this PR for any milestone if it receives feedback only once a month :P

**Comment by TheQwertiest** [comment]
ping?

**Comment by TheQwertiest** [comment]
@jgsogo @memsharded : sorry, but I'm rapidly losing motivation for working on this PR. The last series of feedback ended up in a series of open questions, answers for which are needed to continue the development of this feature. And it's been a month since then (and two months since this PR was submitted).

I understand, that there might be other tasks/priorities that might be preventing you from working on this PR. But a simple "sorry, I don't think we will be able to look into this month/next month/this year/any time soon" would be much better than having zero feedback for a month. Since it just feels that devs are not interested in this PR anymore and actually demotivates the PR submitter from doing any contributions to the project altogether.

Alternatively, if this feature is not something that you want, please, tell me so, so I could close this PR and move on :\\

**Comment by memsharded** [comment]
Hi @TheQwertiest 

I am very sorry we couldn't review this on time for 1.28, and it is being moved for 1.29. You are right, and I understand it can be demotivating, but we do the best we can and work really hard on this project. Even with all the efforts, the amount of activity is so big that it is impossible to keep up with everything, and some things inevitably fail, there might be long delay times to process things, unattended issues, bugs to be fixed that stay there for years… 

Just as a couple of examples of the situation, the team has processed more than **1500 pull requests just in ConanCenter repo (not this Conan repo) https://github.com/conan-io/conan-center-index/pulls** since January, or because of huge demand, we have been doing the online trainings, repeating sessions for 3 months in a row, repeating sessions in EMEA and USA. This takes an incredible amount of time too. 3 months ago we didn't know this was going to happen, and we thought after ConanDays we could focus more on development. Priorities change.

We are definitely interested in this contribution, and moving forward towards a better integration with Meson. Believe me when I say that it is even more frustrating to be on this side, working so hard and still not being able to manage all things that need to be managed as fast as we would like. I really apologize for not being able to put attention to this PR, and I understand if you want to close or abandon it, or you aren't available anymore when we are able to provide feedback. I am still grateful for the contribution, and believe that it will be valuable. 

I am going to propose this plan to move this forward: This might take some time, and as I wouldn’t like to keep the frustration, I will propose that @SSE4 and @jgsogo continue the work. The plan would be the following:

- First, decouple the toolchain from the helper. The goal of the toolchain is that it is possible to call in the command line “meson …” as simple as possible, in the local flow, and still get exactly the same invocation you get with the helper.
- So a first PR would only contain a proposal for a MesonToolchain, that is as simple and minimal as possible. No changes, no new build helper, just do a full command line invocation to Meson, as https://github.com/conan-io/conan/pull/7430 does with Makefiles
- Keep this first PR simple. No need to implement here the whole cross-build functionality, this will probably require more discussion. I’d like to have this merged and released for 1.29
- In a second iteration (a second PR), let's propose the full cross-build functionality. Still no build-helper. Plan would be for Conan 1.30.
- Finally, lets implement the new Meson build helper, in a third PR (target 1.31)

I know this will be slower than desired, but breaking the job in chunks and going step by step will largely increase the likelihood of success, also allowing the team to fully understand the changes going forward. Many thanks again for your contribution

**Comment by TheQwertiest** [comment]
> First, decouple the toolchain from the helper. The goal of the toolchain is that it is possible to call in the command line “meson …” as simple as possible, in the local flow, and still get exactly the same invocation you get with the helper.

That's the exactly the reason for the new helper. I've mentioned this before (https://github.com/conan-io/conan/pull/7155#issuecomment-651355734): it's impossible to reproduce meson invocation via cmd with the old helper, because it uses environment variables to pass various information. So, the toolchain implementation will still have to use the same logic as in this PR, i.e. with your proposed solution there will be two different incompatible implementations all crammed up in the single helper instead of a proper helper with a single implementation.

TLDR: you can't have a reproducible cmd with the old helper.

> Keep this first PR simple. No need to implement here the whole cross-build functionality, this will probably require more discussion. I’d like to have this merged and released for 1.29

This can be done with this PR already: just remove/comment-out all the cross-related parts.

**Comment by SSE4** [comment]
mostly outdated by #8147, #8111 and #7662
I'll be checking if something else useful from this PR might be incorporated

**Comment by memsharded** [comment]
@SSE4 please check when possible if something else needs to be incorporated, or this PR could be closed

**Comment by SSE4** [comment]
@memsharded this could be safely closed. new meson build helper and meson toolchain implement all the things.

**Comment by memsharded** [comment]
This functionality has been splitted and implemented in separate PRs: #8147, #8111 and #7662

It will be released in 1.33.
